File: CSC310-Cpp/_solutions/Module03-Sorting-Lower-Bounds-and-Order-Statistics/debug01/radix_sort.cpp
// radix_sort.cpp
// LSD radix sort on non-negative integers. Each pass is a counting sort
// keyed on one decimal digit (least significant first). Counting sort must
// be stable for the digit passes to combine into a correct overall order.
#include <iostream>
#include <vector>

// Counting sort of a by the digit selected by exp (1, 10, 100, ...).
void countingSortByDigit(std::vector<int>& a, int exp) {
    const int n = static_cast<int>(a.size());
    std::vector<int> output(n);
    std::vector<int> count(10, 0);

    // count[d] = number of elements whose current digit is d
    for (int j = 0; j < n; ++j) {
        ++count[(a[j] / exp) % 10];
    }
    // count[d] = number of elements whose current digit is <= d
    for (int d = 1; d < 10; ++d) {
        count[d] += count[d - 1];
    }
    // Place each element at its final position for this digit, walking
    // from the right so that equal digits keep their relative order.
    for (int j = n - 1; j >= 0; --j) {
        int d = (a[j] / exp) % 10;
        output[count[d] - 1] = a[j];
        --count[d];
    }
    a = output;
}

void radixSort(std::vector<int>& a) {
    int maxValue = 0;
    for (int x : a) {
        if (x > maxValue) {
            maxValue = x;
        }
    }
    for (int exp = 1; maxValue / exp > 0; exp *= 10) {
        countingSortByDigit(a, exp);
        std::cout << "after digit " << exp << ": ";
        for (int x : a) {
            std::cout << x << ' ';
        }
        std::cout << '\n';
    }
}

bool isSorted(const std::vector<int>& a) {
    for (std::size_t i = 1; i < a.size(); ++i) {
        if (a[i - 1] > a[i]) {
            return false;
        }
    }
    return true;
}

int main() {
    std::vector<int> a = {170, 45, 75, 90, 802, 24, 2, 66};
    std::cout << "input:         ";
    for (int x : a) {
        std::cout << x << ' ';
    }
    std::cout << '\n';
    radixSort(a);
    std::cout << (isSorted(a) ? "sorted\n" : "NOT SORTED\n");
    return 0;
}

File: CSC310-Python/Module03-Sorting-Lower-Bounds-and-Order-Statistics/debug01/radix_sort.py
"""LSD radix sort built on counting sort (CLRS Chapter 8).

counting_sort_by_digit sorts a list of non-negative integers by one
decimal digit using the COUNTING-SORT scheme: count occurrences, take
prefix sums, then place each element into the output array.  radix_sort
applies it to the ones digit, then the tens digit, and so on.
"""


def digit(x, place):
    """The decimal digit of x at the given place (0 = ones, 1 = tens, ...)."""
    return (x // 10 ** place) % 10


def counting_sort_by_digit(A, place):
    """Return a new list with A sorted by the digit at 'place'."""
    k = 10
    C = [0] * k
    for x in A:
        C[digit(x, place)] += 1
    # C[d] now holds the number of elements whose digit is <= d.
    for d in range(1, k):
        C[d] += C[d - 1]
    B = [None] * len(A)
    for x in A:
        d = digit(x, place)
        B[C[d] - 1] = x
        C[d] -= 1
    return B


def radix_sort(A, num_digits):
    for place in range(num_digits):
        A = counting_sort_by_digit(A, place)
    return A


def main():
    data = [329, 457, 657, 839, 436, 720, 355]
    print("input :", data)
    print("pass 1:", counting_sort_by_digit(data, 0))
    result = radix_sort(data, 3)
    print("sorted:", result)
    print("correct:", result == sorted(data))

    print()
    ids = [1043, 2087, 1049, 3011, 2043, 1011, 3087, 2011]
    result = radix_sort(ids, 4)
    print("student ids sorted:", result)
    print("correct:", result == sorted(ids))


if __name__ == "__main__":
    main()
